### DISCUSSION TITLE
I'm now learning ANTLR4 for ahk: abstract syntax tree, parser

### DISCUSSION THREAD

**Discussion by FuPeiJiang** [mark-wiemer/ahkpp#34] [General]
I will be using this as reference :D

https://github.com/antlr/grammars-v4/blob/master/javascript/javascript/JavaScriptParser.g4

I hope using this will solve the regex problems

**Comment by mark-wiemer** [comment]
Nice! AST parsing is definitely the better way to format. I want to prioritize automated tests (and other dev QoL improvements) before I rewrite the formatter, but this is a great start!

**Comment by FuPeiJiang** [comment]
they say any serious parser is hand written, I guess it's the most flexible.
if I ever want to write a linter for ahk... idk how to use Antlr for that, because when there are errors, idk how to make the parser work.

And I also saw https://insights.stackoverflow.com/survey/2020#technology-most-loved-dreaded-and-wanted-languages-loved
so I'm learning Rust. Rust can -> web assembly, so this can be used in vscode extension ? I think Perl cannot be used by vscode.
idk yet if it's a hugely bad choice for string manipulation and regex lol. I shall know very soon

coming from ahk, I don't like javascript(async is hard), javascript doesn't have easy string manipulation like python
I would like to pass byref, why copy the value when I don't need to ?
